README.md
# `chronoparse` [![npm version](https://badge.fury.io/js/chronoparse.svg)](https://badge.fury.io/js/chronoparse)[![Build Status](https://travis-ci.com/dylanrenwick/chronoparse.svg?branch=master)](https://travis-ci.com/dylanrenwick/chronoparse)[![Coverage Status](https://coveralls.io/repos/github/dylanrenwick/chronoparse/badge.svg?branch=master)](https://coveralls.io/github/dylanrenwick/chronoparse?branch=master)

chronoparse is a simple nodejs module designed for parsing user-input timespan strings.  
Currently it supports the format:
```
<quantity> <timespan>
```
Where `<quantity>` is a non-negative integer, and `<timespan>` is one of the following:
```
second, seconds, s - 1 second
minute, minutes, m - 60 seconds
hour, hours, h - 60 minutes
day, days, d - 24 hours
week, weeks, w - 7 days
month, months - 30 days
year, years, y - 365 days
decade, decades - 10 years
```

chronoparse parses the given string, and returns an object using the following schema:
```js
{
    totalSeconds, // an integer representing the total length of the timespan in seconds
    startDate, // the given startDate in the options object
    endDate, // a datetime totalSeconds from startDate
    timezone, // the current local timezone, specified as "UTC+/-X" where X is the offset in hours
    parts // an array of each timespan string parsed
}
```

## Usage

chronoparse exports a single function, called `parse`. This function takes a string to be parsed, and an optional options object.

```js
const chronoparse = require("chronoparse");

let result1 = chronoparse("13 days");

console.log(result1.totalSeconds);
// expected output: 1123200

let result2 = chronoparse("2 minutes 3 seconds");

console.log(result2.totalSeconds);
// expected output: 123
```

The options object may contain any or all of the following options:
```js
{
    delimiter, // a string to use to separate multiple timespans, defaults to whitespace
    startDate, // a Date object specifying when to calculate datetimes from, defaults to now
    max // a non-negative integer specifying the maximum number of timespans to parse in the string, defaults to no limit
}
```

### Delimiters

chronoparse can handle compound timespans, such as `2 hours 20 minutes 14 seconds` however, by default, it expects multiple timespans to be separated by only whitespace. If you wish to use other delimiters, such as commas, you should use the `delimiter` option:
```js
let result = chronoparse("2 minutes, 3 seconds", {delimiter: ","});

console.log(result.totalSeconds);
// expected output: 123
```
Note that the trailing space after the delimiter does not need to be specified in the options object, as chronoparse will automatically trim whitespace from the start of the string before attempting to parse the next timespan.

### Max timespans

In some situations, you may wish to only parse a certain number of timespans from the string. In such situations, you should use the `max` option:
```js
let result = chronoparse("1 hour, 2 minutes, 3 seconds", { delimiter: ",", max: 2 });

console.log(result.totalSeconds);
// expected output: 3720
```

chronoparse will parse only the first `max` timespans in a string, and will stop parsing after it has parsed `max` timespans. This means that your string may contain anything, including invalid timespans, after `max` valid timespans, however this is not recommended.
```js
let result = chronoparse("1 hour, 2 minutes, 3 is the coolest number", { delimiter: ",", max: 2 });

console.log(result.totalSeconds);
// expected output: 3720
```

**NOTE:** The default value of `max` is `-1`. This is interpreted as "no limit".

## License and Usage

chronoparse is licensed under the MIT License, meaning it is provided as-is, and can be used freely for any purposes, commercial or otherwise. See [LICENSE](https://github.com/dylanrenwick/chronoparse/blob/master/LICENSE) for the full license.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.0] - 2019-11-18
*** Added
- startDate can now be specified via the options object

## [1.0.0] - 2019-11-18
*** Initial Release

test/chronoparse.text.js
const tzMock = require("timezone-mock");
const chai = require("chai");
const expect = chai.expect;

const parse = require("../src/index.js");

describe("parse", function() {
    const timeStrings = [
        "second",
        "minute",
        "hour",
        "day",
        "week",
        "month",
        "year",
        "decade"
    ];
    const timeSpans = [
        1,
        60,
        60 * 60,
        60 * 60 * 24,
        60 * 60 * 24 * 7,
        60 * 60 * 24 * 30,
        60 * 60 * 24 * 365,
        60 * 60 * 24 * 365 * 10
    ];

    it("should correctly parse simple strings", function() {
        let quantities = [1, 17, 100];
        for(let i in timeStrings) {
            if (!typeof(i) === "number" || i < 0 || i > timeStrings.length) continue;
            for (let j of quantities) {
                let result = timeSpans[i] * j;
                let parseStr = `${j} ${timeStrings[i]}`;
                expect(parse(parseStr).totalSeconds).to.equal(result);
            }
        }
    });

    it("should correctly parse plural names", function() {
        let quantities = [1, 17, 100];
        for(let i in timeStrings) {
            if (!typeof(i) === "number" || i < 0 || i > timeStrings.length) continue;
            for (let j of quantities) {
                let result = timeSpans[i] * j;
                let parseStr = `${j} ${timeStrings[i]}s`;
                expect(parse(parseStr).totalSeconds).to.equal(result);
            }
        }
    });

    it("should correctly parse shortened names", function() {
        let quantities = [1, 17, 100];
        for(let i in timeStrings) {
            if (!typeof(i) === "number" || i < 0 || i > timeStrings.length) continue;
            if (timeStrings[i] === "month" || timeStrings[i] === "decade") continue;
            for (let j of quantities) {
                let result = timeSpans[i] * j;
                let parseStr = `${j}${timeStrings[i][0]}`;
                expect(parse(parseStr).totalSeconds).to.equal(result);
            }
        }
    });

    it("should correctly parse 0 length timespans", function() {
        for(let i in timeStrings) {
            if (!typeof(i) === "number" || i < 0 || i > timeStrings.length) continue;
            let parseStr = `0 ${timeStrings[i]}`;
            expect(parse(parseStr).totalSeconds).to.equal(0);
        }
    });

    it ("should return null for invalid strings", function() {
        expect(parse("no days")).to.be.null;
        expect(parse("0 nones")).to.be.null;
    });

    it ("should correctly parse compound timespans", function() {
        let result = timeSpans[timeStrings.indexOf("week")] * 3 + timeSpans[timeStrings.indexOf("hour")] * 2 + 13;
        expect(parse("3 weeks 2 hours 13 seconds").totalSeconds).to.equal(result);
    });

    it ("should correctly parse compound timespans with custom delimiters", function() {
        let result = timeSpans[timeStrings.indexOf("week")] * 3 + timeSpans[timeStrings.indexOf("hour")] * 2 + 13;
        expect(parse("3 weeks,2 hours,13 seconds", {delimiter: ","}).totalSeconds).to.equal(result);
    });

    it ("should correctly parse compound timespans with custom delimiters and additional whitespace", function() {
        let result = timeSpans[timeStrings.indexOf("week")] * 3 + timeSpans[timeStrings.indexOf("hour")] * 2 + 13;
        expect(parse("3 weeks, 2 hours, 13 seconds", {delimiter: ","}).totalSeconds).to.equal(result);
    });

    it ("should fail to parse compound timespans with unspecified delimiters", function() {
        expect(parse("3 weeks, 2 hours, 13 seconds")).to.be.null;
    });

    it ("should correctly parse compound timespans up to a specified limit", function() {
        let result = timeSpans[timeStrings.indexOf("week")] * 3 + timeSpans[timeStrings.indexOf("hour")] * 2;
        expect(parse("3 weeks 2 hours 13 seconds", {max: 2}).totalSeconds).to.equal(result);
    });

    it ("should correctly parse case insensitively", function() {
        let result = 29
        expect(parse("29 sEconDS", {max: 2}).totalSeconds).to.equal(result);
    });

    it ("should correctly provide the current timezone", function() {
        tzMock.register("US/Eastern");
        expect(parse("1 second").timezone).to.equal("UTC-5");
        tzMock.register("US/Pacific");
        expect(parse("1 second").timezone).to.equal("UTC-8");
        tzMock.register("UTC");
        expect(parse("1 second").timezone).to.equal("UTC+0");
    });

    it ("should return the date of completion as the startDate", function() {
        let now = new Date();
        let result = parse("29 seconds").startDate;
        // Will fail if parsing takes more than 60 seconds
        expect(result - now).to.be.lessThan(60000);
    });

    it ("should return an endDate that matches the given timespan", function() {
        let result = parse("29 seconds");
        expect(result.endDate - result.startDate).to.equal(29000);
    });

    it ("should support specifying a different startDate", function() {
        let date = new Date();
        date.setSeconds(date.getSeconds() + 120);
        let endDate = new Date();
        endDate.setSeconds(endDate.getSeconds() + 149);
        let result = parse("29 seconds", {startDate: date});
        expect(result.startDate.toString()).to.equal(date.toString());
        expect(result.endDate.toString()).to.equal(endDate.toString());
    });
});

src/index.js
const timeStrings = [
    "second",
    "minute",
    "hour",
    "day",
    "week",
    "month",
    "year",
    "decade"
];
const timeSpans = [
    1,
    60,
    60 * 60,
    60 * 60 * 24,
    60 * 60 * 24 * 7,
    60 * 60 * 24 * 30,
    60 * 60 * 24 * 365,
    60 * 60 * 24 * 365 * 10
];
const regex = new RegExp("^(\\d+) ?([a-zA-Z]+)");

function matchToTimespan(match) {
    let quantity = parseInt(match[1]);
    let type = match[2].toLowerCase();
    let typeIndex = timeStrings.indexOf(type);
    if (typeIndex === -1 && type.length === 1) typeIndex = timeStrings.map(x => x[0]).indexOf(type);
    if (typeIndex === -1) typeIndex = timeStrings.map(x => x + "s").indexOf(type);
    if (typeIndex === -1) return -1;
    let timespan = timeTypeToTimespan(typeIndex);
    return timespan === -1 ? timespan : timespan * quantity;
}

function timeTypeToTimespan(type) {
    return timeSpans[type] || -1;
}

function trimString(str, delimiter) {
    while (str.substring(0, delimiter.length) === delimiter) {
        str = str.substring(delimiter.length).trimStart();
    }
    return str;
}

function getTimezone() {
    let offset = new Date().getTimezoneOffset() * -1 / 60;
    return `UTC${offset >= 0 ? "+": ""}${offset}`;
}

module.exports = function parse(text, options = {}) {
    let delimiter = options["delimiter"] || " ";
    let maxCount = options["max"] || -1;
    let allResults = [];
    let result;

    while ((result = regex.exec(text)) !== null && (maxCount === -1 || allResults.length < maxCount)) {
        allResults.push(result);
        text = text.substring(result[0].length);
        text = trimString(text, delimiter);
    }

    if (allResults.length === 0 || ((maxCount < 0 || allResults.length < maxCount) && text.length)) return null;

    let outputData = {
        totalSeconds: 0,
        parts: []
    };
    
    for (let res of allResults) {
        let timespan = matchToTimespan(res);
        if (timespan === -1) return null;
        outputData.totalSeconds += timespan;
        outputData.parts += res[0];
    }

    let date = options.startDate || new Date();
    date.setSeconds(date.getSeconds() + outputData.totalSeconds);
    outputData.startDate = options.startDate || new Date();
    outputData.endDate = date;
    outputData.timezone = getTimezone();

    return outputData;
};